### PR TITLE
Support arbitrary values of the batch axes in {train/test}_on_batch

### PR DESCRIPTION
**Pull request by sadreamer**
This is useful when implementing custom loss functions that require additional data-dependent inputs precomputed somehow outside the graph.

Other related issues (without any answer, unfortunately)
https://github.com/keras-team/keras/issues/8943
https://github.com/keras-team/keras/issues/3660

### CODE REVIEW

**Comment by fchollet** [comment]
> implementing custom loss functions that require additional data-dependent inputs precomputed somehow outside the graph.

This is definitely a niche use case, and in that case I would recommend writing your own version of `model.train_on_batch`. The method is very short (only 7 lines that matter for you), so you can easily write your own.

**Comment by sadreamer** [comment]
@fchollet 

> The method is very short (only 7 lines that matter for you), so you can easily write your own.

The problem is in the method `Model._standardize_user_data`, not in `Model.train_on_batch`. So now I have to copy-paste more than 40 lines of code – does not look like a good idea.
I think when you construct batches by your own and then call `train_on_batch`, you may have an ability to explicitly disable the validation (that is it turned on by default). In my case the only limitation of Keras Model that does not allow to implement the desired loss function ([this one](https://arxiv.org/pdf/1611.00822.pdf)) is the validation that is always turned on. This is sad.

**Comment by fchollet** [comment]
You can just write your own `train_on_batch` and disable validation there. It's just changing a `True` into a `False`, pretty low effort. Bonus is that it will now run faster.

**Comment by sadreamer** [comment]
@fchollet switching `check_batch_axis` to `False` has no effect now since this argument is not used in `Model._standardize_user_data` (seems like a bug btw).

**Comment by fchollet** [comment]
Seems like something to investigate.

**Comment by sadreamer** [comment]
@fchollet so since in the current master `check_batch_axis` is an unused argument of `_standardize_user_data`, it should be removed from the function definition, or I could utilize it by adding the following check: https://github.com/keras-team/keras/pull/9110/files#diff-299cfd5886683a4b012f286403769fc1R1438. Then a workaround for my problem is just to override on my side `train_on_batch` as you propose.
Does this fix work for you? If yes, I could prepare a new PR.
